Introduce support types logic

ShockedPlot7560 opened the pull request:
## Introduction
Many block cannot be placed on certain types of blocks. In addition, there is currently no way for blocks to specify the type of their surface for block placement

## The concept 
It all comes down to the new feature that has been added: ``getSupportType(Facing) : SupportType``. It will then return 4 different types of surfaces according to this table: 
Type | Description
----- | -----
Full | The support face doesn't have any restrictions on support it can provide.
Edge | Support any element which reach the eadges of the block : repeaters,comparator on top of hoppers, cauldron)
Center | Support any elements which can placed in the center (torch on top of fence or wall)
None | No placement is allowed

The discussion and further explanation of this concept can be found here : https://github.com/pmmp/PocketMine-MP/issues/4856

### Relevant issues
* #4856 
* https://github.com/pmmp/PocketMine-MP/issues/458
* https://github.com/pmmp/PocketMine-MP/issues/4529
* https://github.com/pmmp/PocketMine-MP/issues/3299

Some blocks have also had their placement corrected such as ``RedstoneWire`` which was not checking any support

## Changes
### API changes
* Add ``getSupportType()`` in ``Block``
* Add ``SupportType`` classes

### Behavioural changes
Any block that inherits from the parent class : ``Block`` will have a ``None`` surface by default. The default value for ``Opaque`` is: ``Full`` 

## Backwards compatibility
Some block placements have been reviewed and fixed but some may have been impacted due to the change in behavior of the block placement

## Tests
Take the blocks that have been modified and test their placement one by one in all possible positions and situations. 
These were done during development and would be too long to reproduce on video.

ShockedPlot7560 committed:
add SupportType

ShockedPlot7560 committed:
Implement Fence, Opaque and Torch support

ShockedPlot7560 committed:
fix torch placement

ShockedPlot7560 committed:
add Slab placement

ShockedPlot7560 committed:
add PressurePlate, Redstone component placement

ShockedPlot7560 committed:
implement RedstoneWire placement logic

ShockedPlot7560 committed:
add Mushroom, Sapling, SeaLantern placement

ShockedPlot7560 committed:
add stair support logic

ShockedPlot7560 committed:
fix torch placement

ShockedPlot7560 committed:
implement lot of block

ShockedPlot7560 committed:
implement other block

ShockedPlot7560 committed:
fix RedstoneWire placement

ShockedPlot7560 committed:
implement new placement for Rail

ShockedPlot7560 committed:
fix Bed and Beacon placement

ShockedPlot7560 committed:
add Lantern placement

ShockedPlot7560 committed:
fix: bottom supportType for Slab & Stair

ShockedPlot7560 committed:
fix vine placement

ShockedPlot7560 committed:
fix CS

dktapps commented on src/block/Stair.php:
```diff
@@ -128,4 +129,11 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+
+	public function getSupportType(int $facing) : SupportType{
+		if($facing === Facing::UP && $this->isUpsideDown() || $facing === $this->facing || $facing === Facing::DOWN && !$this->isUpsideDown()){
+			return SupportType::FULL();
```
This needs to include logic for corner stairs.

dktapps replied:
This is also potentially subject to precedence issues (needs more ().)

ShockedPlot7560 replied:
in fact, the corner stairs are in this case problematic

ShockedPlot7560 replied:
Do we really implement as vanilla or do we add this feature: 
```
Take a corner stair type of INNER_RIGHT facing north.
In vanilla, you will be able to put torch on the north face and not on the east.
Knowing that east face are full, should we allowed the torch placement or not ?
```

ShockedPlot7560 committed:
fix rail placement

ShockedPlot7560 committed:
add support for corner stairs
- if a face appears as full, the placement will be done

ShockedPlot7560 committed:
prefer Axis comparaison to in_array

ShockedPlot7560 committed:
fix sugarcanne placement
- can be now placed on : dirt / grass / podzol / mycelium / sand

ShockedPlot7560 committed:
fix: farmland support type

ShockedPlot7560 committed:
fix: MonsterSpawner support type
As we can see in Minecraft Bedrock vanilla, that a limited number
of blocks can be placed on the spawner, including Torches.

For a bette coherence, let's assume that the top surface represents
a pillar and the sides would be full.

dktapps commented:
Hi, I haven't had time to review this yet. It's a big PR and I've been busy with other things the last couple of weeks. I'll get to it soon.

ShockedPlot7560 commented:
Ok no problem.

dktapps commented:
A couple of suggestions come to mind (I haven't reviewed the whole thing yet):

`SupportType->hasEdgeSupport()` and `SupportType->hasCenterSupport()` would probably be useful, because:

| Support type | Supports edges | Supports center |
|:--------------:|:------------------:|:------------------:|
| None | | |
| Center | | YES |
| Edge | YES | |
| Full | YES | YES |

Typically this means that we're usually going to be checking for two types of support: `FULL` + one of `CENTER` or `EDGE`.

dktapps committed:
Merge branch 'next-minor' into placement

dktapps commented on src/block/Block.php:
```diff
@@ -642,4 +643,8 @@ public function calculateIntercept(Vector3 $pos1, Vector3 $pos2) : ?RayTraceResu
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		return SupportType::NONE();
```
I'm not sure this really makes sense. `Block` assumes the presence of a 1x1x1 collision cube, which normally equates to full support.

ShockedPlot7560 replied:
Indeed, this part should be as ``Full``, and the ``Transparent`` should be modified for a none support

dktapps reviewed (commented):
Concept is OK, although there's definitely ways to make the code clearer, and some of the choices for support types are a bit weird. Needs more work.

dktapps commented:
Things noticed so far:
- Hopper's upper face seems to be FULL, rather than EDGE. This is counterintuitive, but it's the only way I can explain the behaviour I see (torch can be placed on top of hopper, but not on top of stuff like cauldron or composter, while rails can)
- Spawner is weird. It accepts placement of torches, but not stuff like buttons, levers, pressure plates etc. I don't know what to make of it.
- Buttons should require CENTER support type
- Flower pot should require CENTER support type
- Lanterns should require CENTER support type
- Lever should require CENTER support type
- Torch requires CENTER support when placed on the top face of a block. It can't be placed on top of a cauldron or composter.

... will update via edits while testing

dktapps commented on src/block/DeadBush.php:
```diff
@@ -34,15 +35,15 @@
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$this->getSide(Facing::DOWN)->isTransparent()){
+		if($this->getSide(Facing::DOWN)->getSupportType(Facing::UP)->equals(SupportType::FULL())){
```
I don't think this needs to be changed

dktapps commented on src/block/Lantern.php:
```diff
@@ -77,7 +78,10 @@ protected function canAttachTo(Block $b) : bool{
+		if($face === Facing::DOWN && $blockReplace->getSide(Facing::UP)->getSupportType(Facing::DOWN)->equals(SupportType::NONE())){
+			return false;
+		}
+		if($face === Facing::UP && $blockReplace->getSide(Facing::DOWN)->getSupportType(Facing::UP)->equals(SupportType::NONE())){
```
This can be simplified to check for `Axis::Y` and use `Facing::opposite()` to get the appropriate facings.

dktapps commented on src/block/Hopper.php:
```diff
@@ -99,5 +100,12 @@ public function onScheduledUpdate() : void{
 
+	public function getSupportType(int $facing) : SupportType{
+		if($facing === Facing::UP){
+			return SupportType::EDGE();
```
It's unclear what the correct behaviour of this should be. Many blocks (e.g. torches, lanterns) can be placed on top of a hopper, but can't be placed on top of other EDGE blocks (such as composters); however, rails can be placed on both.

I think the correct thing for hopper would be to treat the top face as FULL. I don't like it, but it seems to be how the game itself treats it.

dktapps commented on src/block/RedMushroom.php:
```diff
@@ -43,7 +44,7 @@ public function onNearbyBlockChange() : void{
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
 		$down = $this->getSide(Facing::DOWN);
-		if(!$down->isTransparent()){
+		if(!$down->getSupportType(Facing::UP)->equals(SupportType::FULL())){
```
I don't think this should be changed

dktapps commented on src/block/Vine.php:
```diff
@@ -116,7 +117,7 @@ protected function recalculateCollisionBoxes() : array{
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$blockClicked->isSolid() || Facing::axis($face) === Axis::Y){
+		if(!$blockClicked->getSupportType($face)->equals(SupportType::FULL()) || Facing::axis($face) === Axis::Y){
```
I don't think this is necessary

dktapps commented:
The following blocks need to be changed:
- Doors (any FULL/EDGE block is sufficient; they should be placeable on slabs, composters and cauldrons)
- Beds (CENTER/EDGE/FULL all seem to be OK)
- Dead bush (these require certain specific blocks, not support types)
- Lantern (the placement logic has not been completely updated)
- Mushroom (the logic has been changed incorrectly; the original logic is correct aside from the missing lightlevel check)

dktapps commented:
- Bell (any support seems to be OK; they can be placed on full blocks, composters, and fences, but not on beds)
- Snow layer (appears to require CENTER support, as best I can tell)
- Coral (appears to require CENTER support)
- Coral fan (appears to require CENTER support)

ShockedPlot7560 commented:
> Spawner is weird. It accepts placement of torches, but not stuff like buttons, levers, pressure plates etc. I don't know what to make of it.

Looking at the wiki and practicing my vanilla tests, we can see that the spawner accepts torches but not other blocks. This can be explained by the fact that it is transparent, and thus everything that is lever, button, pressure plate is useless (does not transmit any current)

dktapps commented:
That's not enough to explain it. Glass has all those same properties, and yet still accepts the placement of all of these blocks.

ShockedPlot7560 commented:
It is true, in this case, it is necessary to leave in a logic external to minecraft vanilla and allow the placement of all

ShockedPlot7560 committed:
add hasCenterSupport && hasEdgeSupport according to review

ShockedPlot7560 committed:
fix support type for Button, Flower pot, Lantern, Lever, Torch
This includes the new hasCenterSupport logic and
factorizes the verification by allocating a dedicated function

The implementation of the lantern is not yet complete

ShockedPlot7560 committed:
remove unnecessary changes

ShockedPlot7560 committed:
fix cs

ShockedPlot7560 committed:
fix cs

ShockedPlot7560 committed:
full implementation of DeadBush

ShockedPlot7560 committed:
change the top face of Hopper to FULL
This change may seem contradictory, however it is also the shortcut that the game seems to make

ShockedPlot7560 committed:
fix door placement

ShockedPlot7560 committed:
fix bed placement + code factorization

ShockedPlot7560 commented:
> Lantern (the placement logic has not been completely updated)

Are you talking about the check to define if the lantern is hanging?
```php
$this->hanging = ($face === Facing::DOWN || !$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->down())));
```

ShockedPlot7560 committed:
implement mushroom light placement

ShockedPlot7560 committed:
change Block and Transparent default support type

ShockedPlot7560 committed:
fix snow, coral placement

ShockedPlot7560 committed:
Button: rename canBeSupportBy to canBeSupportedBy

ShockedPlot7560 committed:
Fence: use Axis comparison instead of Facing

ShockedPlot7560 committed:
Ladder: move placement check in canBeSupportedBy

ShockedPlot7560 committed:
PressurePlate: move placement check in canBeSupportedBy

ShockedPlot7560 committed:
RedstoneComparator: move placement check in canBeSupportedBy

ShockedPlot7560 committed:
RedstoneRepeater: move placement check in canBeSupportedBy

ShockedPlot7560 committed:
RedstoneWire: move placement check in canBeSupportedBy

ShockedPlot7560 commented:
All requested changes have been made. The only things pending, waiting for an answer, for debate are: 
- Spawner: Should we give them a ``FULL`` support for all surfaces, or should we try to stick to minecraft vanilla as currently.
- Lantern: Which investment is currently unimplemented, missing?

dktapps commented on src/block/Transparent.php:
```diff
@@ -23,9 +23,15 @@
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		return SupportType::NONE();
```
This is incorrect. Many transparent blocks are still full cubes.

ShockedPlot7560 replied:
I agree on this point, however, some blocks are not real cube and do not require special support that ``NONE``.  
Instead of defining the supports of the transparent blocks one by one in ``NONE``, it is better to do the case by case of transparent blocks not having a null support ?

dktapps replied:
No, I don't agree. For `Flowable` this would make sense, but not for `Transparent`.

dktapps commented:
> Spawner: Should we give them a FULL support for all surfaces, or should we try to stick to minecraft vanilla as currently.

Spawner shouldn't have any support, since that's how it appears to behave.

dktapps commented on src/block/RedMushroom.php:
```diff
@@ -42,11 +41,18 @@ public function onNearbyBlockChange() : void{
+		$lightLevel = $position->getWorld()->getFullLightAt($position->x, $position->y, $position->z);
+		$id = $block->getId();
+		// TODO: nylium support
+		return ($lightLevel <= 12 && !$block->isTransparent()) || $id === BlockLegacyIds::MYCELIUM || $id === BlockLegacyIds::PODZOL;
```
The light level here will always be zero if the block is opaque.

dktapps replied:
I'm not actually sure this logic belongs in this PR at all. It's unrelated to the objective, which is to introduce support types logic.

ShockedPlot7560 replied:
I'm going to move all this to other pull requests soon if not done before. (Light condition)

dktapps commented:
> Lantern: Which investment is currently unimplemented, missing?

You've left a usage of `canAttachTo()` which has inconsistent behaviour with the new code.

ShockedPlot7560 committed:
add default placement for Flowable to NONE

ShockedPlot7560 committed:
MonsterSpawner: set placement support to NONE

ShockedPlot7560 committed:
Lantern: change for full support type placement

ShockedPlot7560 committed:
RedMushroom: fix light

dktapps commented on src/block/Lantern.php:
```diff
@@ -81,7 +81,7 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
 		}
 
-		$this->hanging = ($face === Facing::DOWN || !$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->down())));
+		$this->hanging = ($face === Facing::DOWN || !$this->canBeSupportedBy($this->position->getWorld()->getBlock($blockReplace->getPosition()->down()), Facing::UP));
```
canAttachTo() is now dead code

dktapps commented on src/block/MonsterSpawner.php:
```diff
@@ -42,11 +40,4 @@ protected function getXpDropAmount() : int{
-			return SupportType::CENTER();
-		}
-		return SupportType::FULL();
-	}
```
This should have NONE support type

ShockedPlot7560 replied:
Indeed, I thought ``Block`` was still setting ``NONE`` by default, so much for me

ShockedPlot7560 committed:
MonsterSpawner: support type to NONE

ShockedPlot7560 committed:
Lantern: remove dead code

ShockedPlot7560 commented:
Is the PhpStan problem important? It's a typing problem, but casting the value as an integer doesn't seem like a good idea.

Looking at the uses of ``getFullLightAt``, I could notice that the use is the same, but no typing error came out.

dktapps commented:
No, it's currently ignored via `actual-problems.neon` baseline

dktapps committed:
Remove unnecessary overrides

dktapps commented on src/block/Fence.php:
```diff
@@ -93,4 +94,11 @@ protected function recalculateCollisionBoxes() : array{
+		if(Facing::axis($facing) === Axis::Y){
+			return SupportType::CENTER();
+		}
+		return SupportType::NONE();
```
This can be reduced to a ternary

dktapps commented on src/block/Hopper.php:
```diff
@@ -99,5 +100,12 @@ public function onScheduledUpdate() : void{
+			return SupportType::FULL();
+		}
+		return SupportType::NONE();
+	}
```
This can be reduced to a ternary

dktapps committed:
more redundant code

dktapps committed:
missed another one

dktapps commented on src/block/SnowLayer.php:
```diff
@@ -81,7 +82,7 @@ protected function recalculateCollisionBoxes() : array{
 
 	private function canBeSupportedBy(Block $b) : bool{
-		return $b->isSolid() || ($b instanceof SnowLayer && $b->isSameType($this) && $b->layers === self::MAX_LAYERS);
+		return $b->getSupportType(Facing::UP)->hasCenterSupport() || ($b instanceof SnowLayer && $b->isSameType($this) && $b->layers === self::MAX_LAYERS);
```
I think the second condition can be removed with this change

dktapps commented on src/block/Torch.php:
```diff
@@ -87,12 +89,21 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
 		return false;
 	}
+
+	protected function canBeSupportBy(Block $support, int $face) : bool{
```
Inconsistent function name

dktapps commented on src/block/Wall.php:
```diff
@@ -76,4 +77,11 @@ protected function recalculateCollisionBoxes() : array{
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		if($facing === Facing::UP || $facing === Facing::DOWN){
```
Facing::axis can be used here

dktapps commented on src/block/Wall.php:
```diff
@@ -76,4 +77,11 @@ protected function recalculateCollisionBoxes() : array{
+			return SupportType::CENTER();
+		}
+		return SupportType::NONE();
+	}
```
This can be reduced to a ternary

dktapps committed:
Adjust support types for some transparent blocks

dktapps commented:
To clarify the reasoning for the inversion of default support types:

Support type differences generally go hand-in-hand with collision box shape differences (with few exceptions). Therefore, since the default behaviour is to assume that blocks have a full collision box, they should also be assumed to have full support on all faces unless otherwise specified. This is why Transparent can't declare NONE support.

I'm aware that this results in more changed files than it would otherwise have done, but this way makes more sense - you'll notice that usually the changes sit right next to `recalculateCollisionBoxes()` or something similar.

dktapps committed:
Update PHPStan baseline

ShockedPlot7560 commented:
That was the behavior I understood in the end. It was not my logic at first, but now it seems more coherent with the rest

ShockedPlot7560 committed:
use ternary + remove dead condition

dktapps commented on src/block/BaseRail.php:
```diff
@@ -220,11 +222,13 @@ private function setConnections(array $connections) : void{
 			foreach($this->getCurrentShapeConnections() as $connection){
-				if(($connection & RailConnectionInfo::FLAG_ASCEND) !== 0 && $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->isTransparent()){
+				$supportType = $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->getSupportType(Facing::UP);
+				if(($connection & RailConnectionInfo::FLAG_ASCEND) !== 0 && ($supportType->equals(SupportType::NONE()) || $supportType->equals(SupportType::CENTER()))){
```
I'm not convinced this is correct. Slope support should depend on the face of the block it leans on, not the top face.

dktapps replied:
`!supportType->hasEdgeSupport()` can be used here

ShockedPlot7560 replied:
It is not even either of the two logics. As can be tested in vanilla, by putting a hopper and rails, the rail seems to be based on the block below it all the time.

dktapps replied:
Never mind, you are correct: 
![image](https://user-images.githubusercontent.com/14214667/169074647-e00cecb8-6057-4071-813a-5425537c6f6c.png)

dktapps commented on src/block/BaseRail.php:
```diff
@@ -38,7 +39,8 @@
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$blockReplace->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $blockReplace->getSide(Facing::DOWN)->getSupportType(Facing::UP);
+		if(!$supportType->equals(SupportType::NONE()) && !$supportType->equals(SupportType::CENTER())){
```
`supportType->hasEdgeSupport()` can be used here

dktapps commented on src/block/BaseRail.php:
```diff
@@ -220,11 +222,13 @@ private function setConnections(array $connections) : void{
 	public function onNearbyBlockChange() : void{
-		if($this->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $this->getSide(Facing::DOWN)->getSupportType(Facing::UP);
+		if($supportType->equals(SupportType::NONE()) || $supportType->equals(SupportType::CENTER())){
```
`!supportType->hasEdgeSupport()` can be used here

dktapps commented on src/block/Slab.php:
```diff
@@ -128,4 +129,13 @@ protected function recalculateCollisionBoxes() : array{
 	}
+
+	public function getSupportType(int $facing) : SupportType{
+		if($this->getSlabType()->equals(SlabType::DOUBLE())){
```
Getter usage unnecessary here

ShockedPlot7560 committed:
Rail: use of hasEdgeSupport && Slab: remove unneccesary getter usage

dktapps commented on src/block/BaseRail.php:
```diff
@@ -38,7 +38,8 @@
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$blockReplace->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $blockReplace->getSide(Facing::DOWN)->getSupportType(Facing::UP);
```
Extra variable can be removed

dktapps commented on src/block/BaseRail.php:
```diff
@@ -220,11 +221,13 @@ private function setConnections(array $connections) : void{
 
 	public function onNearbyBlockChange() : void{
-		if($this->getSide(Facing::DOWN)->isTransparent()){
+		$supportType = $this->getSide(Facing::DOWN)->getSupportType(Facing::UP);
```
same here

dktapps commented on src/block/BaseRail.php:
```diff
@@ -220,11 +221,13 @@ private function setConnections(array $connections) : void{
 		}else{
 			foreach($this->getCurrentShapeConnections() as $connection){
-				if(($connection & RailConnectionInfo::FLAG_ASCEND) !== 0 && $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->isTransparent()){
+				$supportType = $this->getSide($connection & ~RailConnectionInfo::FLAG_ASCEND)->getSupportType(Facing::UP);
```
and here

dktapps commented on src/block/FlowerPot.php:
```diff
@@ -104,19 +104,23 @@ protected function recalculateCollisionBoxes() : array{
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if($this->getSide(Facing::DOWN)->isTransparent()){
+		if($this->canBeSupportedBy($this->getSide(Facing::DOWN), Facing::UP)){
```
```suggestion
		if(!$this->canBeSupportedBy($this->getSide(Facing::DOWN), Facing::UP)){
```

dktapps commented on src/block/Lantern.php:
```diff
@@ -72,22 +73,27 @@ protected function recalculateCollisionBoxes() : array{
-
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		if(!$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->up())) && !$this->canAttachTo($this->position->getWorld()->getBlock($blockReplace->getPosition()->down()))){
+		if(Facing::axis($face) === Axis::Y && !$this->canBeSupportedBy($blockReplace->getSide(Facing::opposite($face)), $face)){
```
This is incorrect. The block should look for up/down supports regardless of which face was clicked.

You can see the problem by right-clicking with a lantern on a wall when there are no available attachment points - the lantern item will be consumed, but no lantern is placed.

dktapps commented on src/block/Sapling.php:
```diff
@@ -125,4 +124,12 @@ private function grow(?Player $player) : bool{
+
+	protected function hasValidSupport() : bool{
+		$down = $this->getSide(Facing::DOWN);
+		if($down->getId() === BlockLegacyIds::GRASS || $down->getId() === BlockLegacyIds::DIRT || $down->getId() === BlockLegacyIds::FARMLAND){
```
Not sure this change belongs in this PR.

dktapps commented on src/block/Sugarcane.php:
```diff
@@ -118,11 +118,24 @@ public function onRandomTick() : void{
+			|| $id === BlockLegacyIds::DIRT
+			|| $id === BlockLegacyIds::PODZOL
+			|| $id === BlockLegacyIds::MYCELIUM
+			|| $id === BlockLegacyIds::SAND;
```
Not sure this change belongs in this PR.

ShockedPlot7560 committed:
Rail: remove extra variable

ShockedPlot7560 committed:
FlowerPot: fi mistake

dktapps commented:
Please don't mark comments as resolved until they've been addressed by commits on the branch

ShockedPlot7560 committed:
remove non related changes

dktapps commented on src/block/DeadBush.php:
```diff
@@ -64,4 +64,12 @@ public function getFlameEncouragement() : int{
+			|| $id === BlockLegacyIds::PODZOL
+			|| $id === BlockLegacyIds::DIRT
+			|| $id === BlockLegacyIds::TERRACOTTA;
+	}
```
This should also be moved to a separate PR

ShockedPlot7560 committed:
Sapling: fix mistake

ShockedPlot7560 committed:
Lantern: fix placement logic

ShockedPlot7560 committed:
DeadBush: remove non related changes

dktapps committed:
tidy
avoid use of protected functions; these signatures arne't consistent or expected to be used in subclasses

dktapps committed:
formatting

dktapps committed:
Merge branch 'next-minor' into placement

dktapps commented:
Updated description:

- this does not fix vines being placeable on the side of cacti - that requires further research
- it does not fix painting interaction with signs
- it does not fix saplings on farmland (that should be done separately)

From what I can see in my testing, this looks pretty much good to go.

ShockedPlot7560 commented:
Thanks for updating, I hadn't noticed that it was obsolete
Once this PR is done, I will make separate placements that have not been fixed by it

dktapps commented on src/block/RedstoneWire.php:
```diff
@@ -45,4 +57,14 @@ public function readStateFromWorld() : void{
+	}
+
+	private function canBeSupportedBy(Block $block) : bool{
+		return $block->getSupportType(Facing::UP)->hasEdgeSupport();
```
```suggestion
		return $block->getSupportType(Facing::UP)->hasCenterSupport();
```

dktapps commented on src/block/Slab.php:
```diff
@@ -125,6 +126,15 @@ protected function recalculateCollisionBoxes() : array{
+	public function getSupportType(int $facing) : SupportType{
+		if($this->slabType->equals(SlabType::DOUBLE())){
+			return SupportType::FULL();
+		}elseif($facing === Facing::UP && $this->slabType->equals(SlabType::TOP()) || $facing === Facing::DOWN && $this->slabType->equals(SlabType::BOTTOM())){
```
```suggestion
		}elseif(($facing === Facing::UP && $this->slabType->equals(SlabType::TOP())) || ($facing === Facing::DOWN && $this->slabType->equals(SlabType::BOTTOM()))){
```

dktapps commented on src/block/SnowLayer.php:
```diff
@@ -80,8 +81,15 @@ protected function recalculateCollisionBoxes() : array{
+
 	private function canBeSupportedBy(Block $b) : bool{
-		return $b->isSolid() || ($b instanceof SnowLayer && $b->isSameType($this) && $b->layers === self::MAX_LAYERS);
+		return $b->getSupportType(Facing::UP)->hasCenterSupport();
```
This requires full support, as far as I can tell from testing

dktapps committed:
missed one

ShockedPlot7560 committed:
Slab: addition of brackets

Co-authored-by: Dylan T. <odigiman@gmail.com>

dktapps committed:
Reduce code

ShockedPlot7560 committed:
RedstoneWire: change SupportType needed to Center

Co-authored-by: Dylan T. <odigiman@gmail.com>

dktapps committed:
Hopper provides down-side center support when funnel points downwards
can support lanterns as seen in vanilla

dktapps committed:
SnowLayer requires full support

dktapps committed:
tidy
